# **Netflix Movies and TV Shows – ETL Project**

## Objectives

- Perform an ETL (Extract, Transform, Load) process on the Netflix dataset.
- Clean and preprocess the data for analysis.
- Handle missing values and format columns appropriately.
- Export the cleaned dataset for future use.

## Inputs
Dataset: `netflix_titles.csv` from Kaggle
Columns include: title, director, cast, country, date_added, release_year, rating, duration, etc.

## Outputs
- Cleaned dataset: `netflix_cleaned.csv`
- Additional column: `year_added`

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

- Missing values are handled using appropriate strategies (e.g., fill with mode or 'Not Specified').
- The `date_added` column is converted to datetime format.
- Whitespaces in string fields are stripped for consistency. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [5]:
import os
current_dir = os.getcwd()
current_dir 

'c:\\Users\\Student\\Downloads\\DET\\Practice PBI\\Power BI\\Netflix-Movies-and-TV-Shows-\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [6]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [7]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Student\\Downloads\\DET\\Practice PBI\\Power BI\\Netflix-Movies-and-TV-Shows-'

In [5]:
!pip install pandas numpy seaborn matplotlib


In [8]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

## Section 1: Extract-Load the Data

Load the dataset

In [ ]:

df = pd.read_csv('data/netflix_titles.csv')

Preview the first few rows

In [ ]:

df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water    Not Specified   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans    Not Specified   
4      s5  TV Show           Kota Factory    Not Specified   

                                                cast        country  \
0                                      Not Specified  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   
3                                      Not Specified  United States   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  
0  As her father nears the end of his life, filmm...      2021.0  
1  After crossing paths at a party, a Cape Town t...      2021.0  
2  To protect his family from a powerful drug lor...      2021.0  
3  Feuds, flirtations and toilet talk go down amo...      2021.0  
4  In a city of coaching centers known to train I...      2021.0

Explore the dataset structure and variables

In [20]:
print("Dataset Shape:", df.shape)
print("\nColumn Names:")
print(df.columns.tolist())
print("\nData Types:")
print(df.dtypes)
print("\nDataset Info:")
print(df.info())

Dataset Shape: (8807, 13)

Column Names:
['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added', 'release_year', 'rating', 'duration', 'listed_in', 'description', 'year_added']

Data Types:
show_id                 object
type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
duration                object
listed_in               object
description             object
year_added             float64
dtype: object

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3 

# Dataset Overview:
* The dataset contains 8,807 records and 13 columns related to Netflix shows and movies.
* Columns include identifiers, metadata (title, director, cast, country), dates (date_added, release_year, year_added),
* categorical data (type, rating, listed_in), and textual descriptions.
* Most columns are of object type (strings), with some numerical columns (release_year, year_added),
* and one datetime column (date_added).
-Note: 'year_added' has some missing values (8709 non-null out of 8807).


## Section 2: Transform: Clean the Data

In [12]:
# Drop rows with missing titles (essential information)
df = df.dropna(subset=['title'])

# Fill missing values
df['director'].fillna('Not Specified', inplace=True)
df['cast'].fillna('Not Specified', inplace=True)
df['country'].fillna(df['country'].mode()[0], inplace=True)
df['rating'].fillna(df['rating'].mode()[0], inplace=True)
df['date_added'].fillna('Unknown', inplace=True)

# Convert 'date_added' to datetime
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

# Create new column for year added
df['year_added'] = df['date_added'].dt.year

# Remove leading/trailing whitespaces from string columns
string_columns = ['type', 'title', 'director', 'cast', 'country', 'rating']
for col in string_columns:
    df[col] = df[col].str.strip()

# Reset index
df.reset_index(drop=True, inplace=True)

# Check for remaining nulls
df.isnull().sum()


show_id          0
type             0
title            0
director         0
cast             0
country          0
date_added      98
release_year     0
rating           0
duration         3
listed_in        0
description      0
year_added      98
dtype: int64

---

# Section 3: Load- Save the Cleaned Dataset

In [19]:
# Save cleaned data in the 'data' folder
df.to_csv('data/netflix_cleaned.csv', index=False)
df.to_excel('data/netflix_cleaned.xlsx', index=False)

print("Cleaned data saved successfully in the 'data' folder.")

Cleaned data saved successfully in the 'data' folder.


Section 2 content

---

NOTE

## Project Notes and Key Findings

### Data Quality Observations:
* The original dataset contained 8,807 records with 12 original columns
* Missing values were present in several key columns (director, cast, country, rating, date_added)
* Successfully handled missing values using appropriate strategies (mode for categorical, "Not Specified" for text)

### ETL Process Summary:
* **Extract**: Loaded Netflix titles dataset from CSV format
* **Transform**: Cleaned missing values, standardized formats, created new `year_added` column
* **Load**: Saved cleaned data in both CSV and Excel formats for future analysis

### Technical Implementation:
* Used pandas for data manipulation and cleaning
* Applied datetime conversion for temporal analysis capabilities
* Implemented data validation and quality checks
* All notebook cells can be run sequentially from top to bottom

### Files Generated:
* `data/netflix_cleaned.csv` - Main cleaned dataset
* `data/netflix_cleaned.xlsx` - Excel version for business users

### Next Steps for Analysis:
* Content type distribution analysis (Movies vs TV Shows)
* Geographic content analysis by country
* Temporal trends in content addition to Netflix
* Rating distribution and content classification analysis
* Duration patterns and viewer preferences

### Data Schema:
The final cleaned dataset contains 13 columns including the new `year_added` field, ready for visualization in Power BI or other analytics tools.

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
